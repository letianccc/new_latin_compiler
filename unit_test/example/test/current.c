void int_const() {
    int i1;
    i1 = 1 < 2 && 2 < 3;
    printf("target 1: %d\n", i1);

    int i2 = 1 < 2 && 2 < 3;
    printf("target 1: %d\n", i2);

    int a = 777;
    int *b = &a;
    printf("target 777: %d\n", a);
    printf("target 777: %d\n", *b);
    *b = 1 < 2 && 2 < 3;
    printf("target 1: %d\n", a);
    printf("target 1: %d\n", *b);
}

void double_const() {
    int i1;
    i1 = 1.1 < 2.1 && 2.1 < 3.1;
    printf("target 1: %d\n", i1);

    int i2 = 1.1 < 2.1 && 2.1 < 3.1;
    printf("target 1: %d\n", i2);

    int a = 777;
    int *b = &a;
    printf("target 777: %d\n", a);
    printf("target 777: %d\n", *b);
    *b = 1.1 < 2.1 && 2.1 < 3.1;
    printf("target 1: %d\n", a);
    printf("target 1: %d\n", *b);
}

void double_mix() {
    int i1;
    double d1 = 1.1;
    double d2 = 2.1;
    double d3 = 3.1;

    i1 = 777;
    i1 = 1.1 < d2 && d2 < 3.1;
    printf("target 1: %d\n", i1);

    i1 = 777;
    i1 = 1.1 < d2 || d2 < 3.1;
    printf("target 1: %d\n", i1);

    i1 = 777;
    i1 = 1.1 > d2 || d2 > 3.1;
    printf("target 0: %d\n", i1);

    i1 = 777;
    i1 = d1 < 2.1 && 2.1 < d3;
    printf("target 1: %d\n", i1);

    i1 = 777;
    i1 = d1 < d2 && d2 < d3;
    printf("target 1: %d\n", i1);

    int a = 777;
    int *b = &a;
    printf("target 777: %d\n", a);
    printf("target 777: %d\n", *b);
    *b = 1.1 < d2 && d2 < 3.1;
    printf("target 1: %d\n", a);
    printf("target 1: %d\n", *b);
}
