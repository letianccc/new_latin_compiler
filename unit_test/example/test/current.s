//by latin

    .text
    .globl	_pass_pointer
_pass_pointer:
    pushl	%ebp
    pushl	%ebx
    pushl	%esi
    pushl	%edi
    movl	%esp, %ebp
    andl	$-16, %esp
    subl	$12, %esp
    movl	20(%ebp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 8(%esp)
    movl	8(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC0, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	%ebp, %esp
    popl	%edi
    popl	%esi
    popl	%ebx
    popl	%ebp
    ret
    .text
    .globl	_mul_indirect_assign
_mul_indirect_assign:
    pushl	%ebp
    pushl	%ebx
    pushl	%esi
    pushl	%edi
    movl	%esp, %ebp
    andl	$-16, %esp
    subl	$60, %esp
    movl	$1, %eax
    movl	%eax, 56(%esp)
    leal	56(%esp), %eax
    movl	%eax, 44(%esp)
    movl	44(%esp), %eax
    movl	%eax, 52(%esp)
    leal	52(%esp), %eax
    movl	%eax, 40(%esp)
    movl	40(%esp), %eax
    movl	%eax, 48(%esp)
    movl	56(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC0, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	52(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 36(%esp)
    movl	36(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC0, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	48(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 32(%esp)
    movl	32(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 28(%esp)
    movl	28(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC0, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	48(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 24(%esp)
    movl	24(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 20(%esp)
    movl	24(%esp), %eax
    movl	$2, %edx
    movl	%edx, (%eax)
    movl	56(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC1, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	52(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 16(%esp)
    movl	16(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC1, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	48(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 12(%esp)
    movl	12(%esp), %eax
    movl	%eax, %eax
    movl	(%eax), %eax
    movl	%eax, 8(%esp)
    movl	8(%esp), %eax
    movl	%eax, 4(%esp)
    movl	$LC1, %eax
    movl	%eax, 0(%esp)
    call	_printf
    movl	%ebp, %esp
    popl	%edi
    popl	%esi
    popl	%ebx
    popl	%ebp
    ret
    .text
    .globl	_main
_main:
    pushl	%ebp
    pushl	%ebx
    pushl	%esi
    pushl	%edi
    movl	%esp, %ebp
    andl	$-16, %esp
    subl	$16, %esp
    movl	$1, %eax
    movl	%eax, 12(%esp)
    leal	12(%esp), %eax
    movl	%eax, 4(%esp)
    movl	4(%esp), %eax
    movl	%eax, 8(%esp)
    movl	8(%esp), %eax
    movl	%eax, 0(%esp)
    call	_pass_pointer
    call	_mul_indirect_assign
    call	_getchar
    movl	%ebp, %esp
    popl	%edi
    popl	%esi
    popl	%ebx
    popl	%ebp
    ret
LC0:
    .string	"target 1: %d\n"
LC1:
    .string	"target 2: %d\n"
