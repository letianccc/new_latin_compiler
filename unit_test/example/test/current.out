target 3: 3
target 3: 3
target 3: 3
target 3: 3
target 3: 3
target 3: 3
target 3: 3
target 3: 3
target 3: 3
